libsplinter/src/service/scabbard/state.rs

```diff
@@ -284,22 +275,67 @@ impl ScabbardState {
         &mut self.batch_history
     }
 
-    pub fn subscribe_to_state(&mut self, request: Request) -> Result<Response, ResponseError> {
-        let events = self.commit_history.events()?;
-        self.event_dealer
-            .subscribe(request, &mut events.into_iter())
+    pub fn subscribe_to_state(
+        &mut self,
+        request: Request,
+        last_txn_receipt: Option<String>,
+    ) -> Result<Response, ResponseError> {
+        let all_receipts = self.transaction_receipt_store.iter().map_err(|err| {
+            ResponseError::CatchUpHistoryError(format!(
+                "failed to get transaction receipts from store: {}",
+                err
+            ))
+        })?;
+
+        let unseen_receipts = if let Some(txn_id) = last_txn_receipt {
+            all_receipts
+                .skip_while(|receipt| receipt.transaction_id != txn_id)
+                .skip(1)
+                .collect::<Vec<_>>()
```

Comment:
won't this potentially load an unbounded amount of stuff into memory?

Reply:
Potentially, yes. Couple things to note, though: 1) the `TransactionReceiptStore.iter()` method is already doing that, because it returns an iterator over owned items. 2) there are plans to "age-out" old transaction receipts when they aren't necessary to reproduce the current state, so this will be bounded by the amount of addresses that are used by the smart contracts.

Reply:
Improvements could be made by adding a `range_iter` method, that can be translated by the underlying store implementation into more efficient query.  Those changes would need to be made in `libsawtooth` first.

Either way, you could avoid the `collect` call entirely and just use the iterators.

Reply:
@jsmitchell Iterators are lazy.

Reply:
I think we could use the iterators directly here. The ordered store implementations in libsawtooth do load all items into memory as they're currently implemented, but that could be changed in the future.

Reply:
That should probably be changed pretty quickly...

Reply:
Updated the "Send past state deltas to scabbard subscriber" commit to use the iterators directly.